components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodTerminalService.kt

```diff
@@ -73,24 +90,139 @@ class GitpodTerminalService(private val project: Project) : Disposable {
 
         messageBusConnection.subscribe(ToolWindowManagerListener.TOPIC, toolWindowManagerListener)
 
-        return completableFuture
+        completableFuture.get()
+    }
+
+    private tailrec fun waitUntilGuestClientIsConnected() {
```

Comment:
I wonder should not we rather merge this class into `GitpodClientProjectSessionTracker`? and rename to `GitpodClientProjectSessionManager`? 

It has a project and a session, so no need for guessing?

Reply:
It's possible, indeed. There's just one difference, as far as I remember:
As `GitpodClientProjectSessionTracker` is marked with `client="guest"`, the class is being invoked everytime a client connects or reconnects.
While `GitpodTerminalService` currently runs only when the backend starts.
So it's necessary to add a flag to determine if the shared terminals have already been created and don't create them again in case of reconnection.

```
        <projectService serviceImplementation="io.gitpod.jetbrains.remote.GitpodClientProjectSessionTracker" client="guest" preload="true"/>
        <projectService serviceImplementation="io.gitpod.jetbrains.remote.GitpodTerminalService" preload="true"/>
```

But wouldn't it be better to do it in a follow-up PR as it's a refactoring?

Reply:
ok, let's do later if we have problems with multiple clients connecting